src/Data/Singletons/Deriving/Show.hs

```diff
@@ -0,0 +1,140 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module      :  Data.Singletons.Deriving.Show
+-- Copyright   :  (C) 2015 Richard Eisenberg
+-- License     :  BSD-style (see LICENSE)
+-- Maintainer  :  Richard Eisenberg (rae@cs.brynmawr.edu)
+-- Stability   :  experimental
+-- Portability :  non-portable
+--
+-- Implements deriving of Show instances
+--
+----------------------------------------------------------------------------
+{-# LANGUAGE ScopedTypeVariables #-}
+module Data.Singletons.Deriving.Show (mkShowInstance) where
+
+import Language.Haskell.TH.Syntax hiding (showName)
+import Language.Haskell.TH.Ppr (pprint)
+import Language.Haskell.TH.Desugar
+import Data.Singletons.Names
+import Data.Singletons.Util
+import Data.Singletons.Syntax
+import Data.Singletons.Deriving.Infer
+import Control.Monad (when)
+import Data.Maybe (fromMaybe)
+import GHC.Lexeme (startsConSym, startsVarSym)
+import GHC.Show (appPrec, appPrec1)
+
+mkShowInstance :: DsMonad q => DType -> [DCon] -> q UInstDecl
+mkShowInstance ty cons = do
+  when (null cons) $
```

Comment:
Well, standalone deriving _can_ handle it, but it's perhaps not the implementation you'd expect:

```
λ> data Foo
λ> deriving instance Show Foo

==================== Derived instances ====================
Derived instances:
  instance GHC.Show.Show Ghci4.Foo where
    GHC.Show.showsPrec = GHC.Err.error "Void showsPrec"
```

We don't have the ability to promote `error` yet, so that's not an option. And I'm not sure what other implementation you have in mind for empty datatypes, so I left it as an error.